File: dotnet/Bosbec.CreateSpecificationExample/src/Bosbec.CreateSpecificationExample/Helpers/AuthenticationHelper.cs
namespace Bosbec.CreateSpecificationExample.Helpers
{
    using System;
    using System.Net;

    using EasyHttp.Http;

    public static class AuthenticationHelper
    {
        public static string Authenticate(string username, string password)
        {
            var baseUrl = ConfigurationHelper.ApiBaseUrl;

            if (string.IsNullOrEmpty(baseUrl))
            {
                throw new InvalidOperationException("Missing MobileResponse API base url in configuration file");
            }

            var client = new HttpClient(baseUrl);

            var data = new
            {
                username,
                password
            };

            var response = client.Post("api/v2/authentication-tokens", data, "application/json");

            if (response.StatusCode != HttpStatusCode.OK)
            {
                return null;
            }

            var authenticationToken = (string)response.DynamicBody.id;

            return authenticationToken;
        }

        public static string Authenticate()
        {
            return Authenticate(ConfigurationHelper.Username, ConfigurationHelper.Password);
        }
    }
}

File: java/Bosbec.CreateSpecificationExample/src/main/java/org/bosbec/createspecificationexample/AuthenticationHelper.java
package org.bosbec.createspecificationexample;

import org.apache.commons.io.IOUtils;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.ContentType;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.HttpClients;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;

public class AuthenticationHelper {
    public static String authenticate(String username, String password) {
        try {
            URL baseUri = new URL(ConfigurationHelper.getApiBaseUrl());

            URL url = new URL(baseUri, "api/v2/authentication-tokens");

            HttpPost request = new HttpPost(url.toString());

            ContentType contentType = ContentType.create("application/json", "UTF-8");

            String body = new JSONObject()
                    .put("username", username)
                    .put("password", password)
                    .toString();

            StringEntity content = new StringEntity(body, contentType);

            request.setEntity(content);

            HttpClient client = HttpClients.createDefault();

            HttpResponse response = client.execute(request);

            String json = IOUtils.toString(new InputStreamReader(response.getEntity().getContent()));

            JSONObject authenticationToken = new JSONObject(json);

            return authenticationToken.getString("id");
        } catch (IOException e) {
            e.printStackTrace();
        }

        return null;
    }

    public static String authenticate() {
        String username = "kim@bosbec.se";
        String password = "kalle";

        return authenticate(username, password);
    }
}